File: src-terminal/com/jediterm/terminal/ArrayTerminalDataStream.java
package com.jediterm.terminal;

import java.io.IOException;

/**
 * Takes data from underlying char array.
 *
 * @author traff
 */
public class ArrayTerminalDataStream implements TerminalDataStream {
  protected final char[] myBuf;
  protected int myOffset;
  protected int myLength;

  public ArrayTerminalDataStream(char[] buf, int offset, int length) {
    myBuf = buf;
    myOffset = offset;
    myLength = length;
  }

  public ArrayTerminalDataStream(char[] buf) {
    this(buf, 0, buf.length);
  }

  public char getChar() throws IOException {
    if (myLength == 0) {
      throw new EOF();
    }

    myLength--;

    return myBuf[myOffset++];
  }

  public void pushChar(final char b) throws EOF {
    if (myOffset == 0) {
      // Pushed back too many... shift it up to the end.
      myOffset = myBuf.length - myLength;
      System.arraycopy(myBuf, 0, myBuf, myOffset, myLength);
    }

    myLength++;
    myBuf[--myOffset] = b;
  }

  public String readNonControlCharacters(int maxChars) throws IOException {
    String nonControlCharacters = CharacterUtils.getNonControlCharacters(maxChars, myBuf, myOffset, myLength);

    myOffset += nonControlCharacters.length();
    myLength -= nonControlCharacters.length();

    return nonControlCharacters;
  }

  public void pushBackBuffer(final char[] bytes, final int length) throws EOF {
    for (int i = length - 1; i >= 0; i--) {
      pushChar(bytes[i]);
    }
  }
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/ArrayTerminalDataStream.kt
package com.jediterm.terminal

import com.jediterm.terminal.util.CharUtils
import java.io.IOException

/**
 * Takes data from underlying char array.
 *
 * @author traff
 */
open class ArrayTerminalDataStream @JvmOverloads constructor(
    protected var myBuf: CharArray,
    protected var myOffset: Int = 0,
    protected var myLength: Int = myBuf.size - myOffset
) : TerminalDataStream {
    @get:Throws(IOException::class)
    override val char: Char
        get() {
            if (myLength == 0) {
                throw TerminalDataStream.EOF()
            }

            myLength--

            return myBuf[myOffset++]
        }

    @Throws(TerminalDataStream.EOF::class)
    override fun pushChar(c: Char) {
        if (myOffset == 0) {
            // Pushed back too many... shift it up to the end.

            val newBuf: CharArray
            if (myBuf.size - myLength == 0) {
                newBuf = CharArray(myBuf.size + 1)
            } else {
                newBuf = myBuf
            }
            myOffset = newBuf.size - myLength
            System.arraycopy(myBuf, 0, newBuf, myOffset, myLength)
            myBuf = newBuf
        }

        myLength++
        myBuf[--myOffset] = c
    }

    @Throws(IOException::class)
    override fun readNonControlCharacters(maxChars: Int): String? {
        val nonControlCharacters = CharUtils.getNonControlCharacters(maxChars, myBuf, myOffset, myLength)

        myOffset += nonControlCharacters.length
        myLength -= nonControlCharacters.length

        return nonControlCharacters
    }

    @Throws(TerminalDataStream.EOF::class)
    override fun pushBackBuffer(bytes: CharArray?, length: Int) {
        bytes?.let {
            for (i in length - 1 downTo 0) {
                pushChar(it[i])
            }
        }
    }

    override val isEmpty: Boolean
        get() = myLength == 0
}
